docs/design/sparse-v2.md

```diff
@@ -0,0 +1,145 @@
+# Sparse Patterns v2 redeisgn
+
+Authors: [Daniel Ploch](mailto:dploch@google.com)
+
+**Summary:** This Document documents a redesign of the sparse patterns list
+stored in the JJ working copy, in order to facilitate several desirable
+improvements. It covers both the migration path and the planned end state.
+
+## Current State
+
+Sparse patterns are encoded as a list of prefix strings in the working copy
+proto and as a list of RepoPaths in Rust code. Patterns can be added to or
+removed from this list, and the list can be converted into a Matcher which
+visits only paths matching at least one prefix in the list. Additionally,
+the sparse set is ephemeral, tracked only in the working copy proto. We aim
+to improve Sparse Patterns through several steps.
+
+## Redesign Overview
+
+The redesign (hereafter "Sparse Patterns v2") aims to accomplish three goals:
+
+1.  Sparse Patterns should be versioned with the working copy
+1.  Sparse Patterns should support more flexible matching rules
+1.  Sparse Patterns should support client path remapping
+
+The ability to support sub-exclusions and non-recursive inclusions is vital to
+working with large mono repos with gigantic subtrees. Sometimes it is necessary
+to include a relevant parent directory with code, but exclude giant
+subdirectories with test data, images, i18n, etc.  Non-recursive inclusions are
+vital for very wide directories.
+
+With the added complexity and the potential for custom integrations to go wrong
+for different pattern sets, versioning the Sparse Patterns set becomes very
+useful as well.
+
+### Versioning and Storage
+
+The SparsePatterns [proto](https://github.com/martinvonz/jj/blob/066f31a15d6605c596928ebd8b01096120f66b7d/lib/src/protos/working_copy.proto#L35)
+will first have its definition moved into the op_store for persistent storage.
+Like all other objects, it will gain a content-based hash id for unique
+reference. For performance, the default 'everything' SparsePattern (a list
+containing one element, the empty string) will be special-cased to avoid storage
+penalties. The OpStore trait will gain read+write methods for SparsePatterns
+objects, that function as expected.
+
+The [View](https://github.com/martinvonz/jj/blob/066f31a15d6605c596928ebd8b01096120f66b7d/lib/src/protos/op_store.proto#L61)
+object will change to store structured data for each working copy:
+
+```
```

Comment:
Nit: specify the language for the codeblock.

Reply:
Done.